### PR TITLE
Refactor sorting by CreationTimestamp

### PR DESCRIPTION
**Pull request by rhcarvalho**
Cleaning up multiple implementations of sorting builds by `CreationTimestamp`

@mfojtik @bparees PTAL

### CODE REVIEW

File: pkg/cmd/cli/describe/describer.go [hunk 1]
```diff
@@ -361,11 +345,8 @@ func (d *BuildConfigDescriber) Describe(namespace, name string) (string, error)
 			return nil
 		}
 		fmt.Fprintf(out, "Builds:\n  Name\tStatus\tDuration\tCreation Time\n")
-		sortedBuilds := sortableBuilds(builds.Items)
-		sort.Sort(sortedBuilds)
-		for i := range sortedBuilds {
-			// iterate backwards so we're printing the newest items first
```
Comment (rhcarvalho) [inline pkg/cmd/cli/describe/describer.go]: No need to iterate backwards if we sort the items properly in the first place.

**Comment by mfojtik** [comment]
LGTM [test]

File: pkg/cmd/cli/describe/describer.go [hunk 2]
```diff
@@ -361,11 +345,8 @@ func (d *BuildConfigDescriber) Describe(namespace, name string) (string, error)
 			return nil
 		}
 		fmt.Fprintf(out, "Builds:\n  Name\tStatus\tDuration\tCreation Time\n")
-		sortedBuilds := sortableBuilds(builds.Items)
```
Comment (rhcarvalho) [inline pkg/cmd/cli/describe/describer.go]: We don't need to declare a new variable here. It is a slice, the contents are not copied but just referenced. Changing (sorting) `sortedBuilds` also affects `builds.Items`. http://play.golang.org/p/Tk0FbAvk_W

File: pkg/build/prune/resolvers.go [hunk 3]
```diff
@@ -98,30 +98,34 @@ func (o *perBuildConfigResolver) Resolve() ([]*buildapi.Build, error) {
...
-				completeBuilds = append(completeBuilds, build)
+				completeBuilds = append(completeBuilds, *build)
 			} else if failedStates.Has(string(build.Status.Phase)) {
-				failedBuilds = append(failedBuilds, build)
+				failedBuilds = append(failedBuilds, *build)
 			}
 		}
-		sort.Sort(sortableBuilds(completeBuilds))
-		sort.Sort(sortableBuilds(failedBuilds))
+		sort.Sort(sort.Reverse(buildapi.ByCreationTimestamp(completeBuilds)))
```
Comment (liggitt) [inline pkg/build/prune/resolvers.go]: all of these uses assume `buildapi.ByCreationTimestamp` sorts in place... that's probably not a great assumption

Comment (rhcarvalho) [inline pkg/build/prune/resolvers.go]: `buildapi.ByCreationTimestamp` is a type, we're type casting there. `sort.Sort` sorts in place.

File: pkg/build/prune/resolvers.go [hunk 4]
```diff
@@ -98,30 +98,34 @@ func (o *perBuildConfigResolver) Resolve() ([]*buildapi.Build, error) {
 	completeStates := util.NewStringSet(string(buildapi.BuildPhaseComplete))
 	failedStates := util.NewStringSet(string(buildapi.BuildPhaseFailed), string(buildapi.BuildPhaseError), string(buildapi.BuildPhaseCancelled))
 
-	results := []*buildapi.Build{}
+	prunableBuilds := []buildapi.Build{}
```
Comment (bparees) [inline pkg/build/prune/resolvers.go]: why can't this be an array of pointers?  there's going to be a lot more copying going on now right?

**Comment by rhcarvalho** [comment]
@bparees I will rework this later.

File: pkg/build/prune/resolvers.go (hunk 4 above)
Comment (rhcarvalho) [inline pkg/build/prune/resolvers.go]: You're right. I went originally with `ByCreationTimestamp` as `[]Build` following the definition of `BuildList`:
```go
// BuildList is a collection of Builds.
type BuildList struct {
	kapi.TypeMeta
	kapi.ListMeta

	// Items is a list of builds
	Items []Build
}
```

It's more reasonable to have `ByCreationTimestamp` as `[]*Build` and do explicity type conversions where needed.

**Comment by rhcarvalho** [comment]
@bparees I've done what I wanted to rework. Instead of defining sort on `[]Build`, now we're always sorting on `[]*Build`. AFAICT converting the former into the latter where needed has no negative impact on performance (taking the address of existing `Build`s -- O(n), versus  O(n*log(n)) for `sort.Sort`), and avoids copying during sort.

**Comment by smarterclayton** [comment]
Cache locality is a far bigger driver on sort performance than copy.  Copy is at least an order of magnitude faster than missing the cache.  In general, assuming that []*Build is faster is something to prove, not assume.

**Comment by bparees** [comment]
@smarterclayton https://gist.github.com/bparees/24b8630e2daf6a54bbb2

it's admittedly a trivial benchmark (no warming, among other things) but from what i'm seeing as soon as you get beyond trivial numbers of items and trivially sized objects, pointer sort performance is better (in fact in these runs it looks like value performance was *never* better).  which doesn't surprise me that much, personally.  it's been a while since i've done computer architecture, but i'm guessing copying the values around is actually causing cache misses (because you're filling the cache w/ content you don't really need) which hurts you more than the locality of being able to fill cache lines with multiple colocated values from memory.

not to mention who knows what other useful cache content (for other threads) you pushed out of the cache while copying those values around for the sort.

**Comment by smarterclayton** [comment]
In this case your sizes are enough that they are probably still in L3, and
you are initializing the objects in close proximity to the sort (which
means you haven't had a chance to have your working set be moved to main
memory).  I don't think these tests are sized sufficiently to reflect the
chance of missing L3 sufficiently.

Also, the objects you are comparing are only 128 bits wide, and you're not
sorting by the string (which itself is a ref).  Whats the outcome when
sorting something much broader (like Build which is at least 20 pointers
wide), and doing it based on the string?

Also, random and linear access on the []*Blog is going to be worse than
[]Blog, so if you are passing this to a caller who has to traverse it then
you're still better off as a []Blog in many cases.  I don't know that a
single pass is enough to recoup the cost but you'll want to preserve the
cache accesses for the strings (things you are printing).

On Aug 8, 2015, at 10:46 AM, Ben Parees <notifications@github.com> wrote:

@smarterclayton <https://github.com/smarterclayton>
https://gist.github.com/bparees/24b8630e2daf6a54bbb2

it's admittedly a trivial benchmark (no warming, among other things) but
from what i'm seeing as soon as you get beyond trivial numbers of items and
trivially sized objects, pointer sort performance is better (in fact in
these runs it looks like value performance was *never* better). which
doesn't surprise me that much, personally. it's been a while since i've
done computer architecture, but i'm guessing copying the values around is
actually causing cache misses (because you're filling the cache w/ content
you don't really need) which hurts you more than the locality of being able
to fill cache lines with multiple colocated values from memory.

not to mention who knows what other useful cache content (for other
threads) you pushed out of the cache while copying those values around for
the sort.

—
Reply to this email directly or view it on GitHub
<https://github.com/openshift/origin/pull/4043#issuecomment-128987847>.

**Comment by bparees** [comment]
how about: https://gist.github.com/bparees/d27e0e651289e34ad67d

haswell l3 caches are 8 megs: http://www.7-cpu.com/cpu/Haswell.html

i'm generating 10k items of size 10k (so ~100megs) of garbage data twice, after generating my sortable data and before running the benchmarked sorts.  i'm also sorting that data and evaluating the result of that sort to hopefully avoid the optimizer ripping all that code out.

**Comment by bparees** [comment]
ugh, ignore, bad variable naming,

**Comment by bparees** [comment]
fixed: https://gist.github.com/bparees/6be4ae709fc036b515a4

10 loops of generating 20k items with 10k of string data each.  each 10k set of items gets sorted (one as pointers, one as values for 20k total)

results are running now, will probably take a while.

**Comment by bparees** [comment]
(no i don't have anything better to do today)

**Comment by smarterclayton** [comment]
The best part about computers is that they are completely arbitrary now.

On Aug 8, 2015, at 11:33 AM, Ben Parees <notifications@github.com> wrote:

(no i don't have anything better to do today)

—
Reply to this email directly or view it on GitHub
<https://github.com/openshift/origin/pull/4043#issuecomment-128990953>.

**Comment by bparees** [comment]
preliminary results:
https://gist.github.com/bparees/0dd92dff472e25e15869

**Comment by bparees** [comment]
final results:
https://gist.github.com/bparees/2da292017933945ba3d2

value was faster in 18 of 560 cases

**Comment by bparees** [comment]
it occurs to me that given my allocation strategy (allocate all the items one after another), it's possible even i the pointer case that all the values exhibit good locality, given that the memory is unlikely to be significantly fragmented.

**Comment by smarterclayton** [comment]
Seems likely.  A second order effect these benchmarks don't track - you
have a *lot* more garbage for the GC to track when you use pointers.  It's
likely we would veto pointers in any big array for specifically that reason
- sort would have to be substantially faster (an order of magnitude?) to
justify all the garbage you're creating (and the memory needed to track all
of those individual pointers).

On Aug 9, 2015, at 11:06 AM, Ben Parees <notifications@github.com> wrote:

it occurs to me that given my allocation strategy (allocate all the items
one after another), it's possible even i the pointer case that all the
values exhibit good locality, given that the memory is unlikely to be
significantly fragmented.

—
Reply to this email directly or view it on GitHub
<https://github.com/openshift/origin/pull/4043#issuecomment-129196413>.

**Comment by bparees** [comment]
I feel like you're overstating the case there. It's one pointer per item,
the items already contain pointers and multiple objects. So you're (far)
less than doubling the object count when you introduce the pointer. And by
copying pointers around instead of objects you are allocating/churning less
volume of memory (unless go has a direct swap operation that doesn't create
a temporary object)

In anycase it's all short lived objects which should be well handled by any
decently optimized gc algorithm.

Also it's an array of pointers. Shouldn't that be gc'd as a single object,
effectively, just like the array of structs?
I may attempt a striped/randomized allocation benchmark later.

Ben Parees | OpenShift

**Comment by smarterclayton** [comment]
On Aug 9, 2015, at 1:47 PM, Ben Parees <notifications@github.com> wrote:

I feel like you're overstating the case there. It's one pointer per item,
the items already contain pointers and multiple objects. So you're (far)
less than doubling the object count when you introduce the pointer. And by
copying pointers around instead of objects you are allocating/churning less
volume of memory (unless go has a direct swap operation that doesn't create
a temporary object)


I'm talking about normal use.  10k calls to allocate 100 bytes creates 10k
items in GC vs one allocation of 1M is hard to argue is the right
approach.  When you swap objects in sort in go you pay for memmove but
that's it - you're not allocating any memory during a sort in either case
(unless your sort is wrong).  Strings are also special, and that's the bulk
of the object structure.  Making a sort faster is just one part of the
overall story, and it's not something you really need to optimize for in
any of the use cases we have on the server side (memory allocation is
almost always the worse evil).


In anycase it's all short lived objects which should be well handled by any
decently optimized gc algorithm.



You're moving costs around (from sort to gc) though by switching to the
pointer.  I'm just pointing out that I would usually weigh allocation costs
and patterns that encourage that >> sort cost (a one time thing in most
flows).  For anything that is not sorted, it's a no brainer.

Every object in GC has to be traversed during a walk until Go has
generations, and even then the generation structure has to be traversed for
promotion out of the nursery, so it's not a free operation to put more
pressure on GC.  Put another way, how expensive is it for you to build a
[]*Build from a []Build (one allocate and one traversal) and the sort is
then faster, so we both win.  The converse is not true.



Also it's an array of pointers. Shouldn't that be gc'd as a single object,
effectively, just like the array of structs?
I may attempt a striped/randomized allocation benchmark later.



I'm talking about the sub objects *Build which are individually allocated
vs all at once.  In fact, that's something I should look at in the storage
code that we are decoding into the existing object in the list, vs creating
and copying.



Ben Parees | OpenShift
On Aug 9, 2015 12:21, "Clayton Coleman" <notifications@github.com> wrote:

> Seems likely. A second order effect these benchmarks don't track - you
> have a *lot* more garbage for the GC to track when you use pointers. It's
> likely we would veto pointers in any big array for specifically that
reason
> - sort would have to be substantially faster (an order of magnitude?) to
> justify all the garbage you're creating (and the memory needed to track
all
> of those individual pointers).
>
> On Aug 9, 2015, at 11:06 AM, Ben Parees <notifications@github.com> wrote:
>
> it occurs to me that given my allocation strategy (allocate all the items
> one after another), it's possible even i the pointer case that all the
> values exhibit good locality, given that the memory is unlikely to be
> significantly fragmented.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/4043#issuecomment-129196413>.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/4043#issuecomment-129212433>.
>

—
Reply to this email directly or view it on GitHub
<https://github.com/openshift/origin/pull/4043#issuecomment-129221661>.

**Comment by bparees** [comment]
> I'm talking about normal use. 10k calls to allocate 100 bytes creates 10k items in GC vs one allocation of 1M is hard to argue is the right approach. 

but it's not one allocation in the other case because you still have to allocate the build objects you retrieved, so you're talking about 10k+1 allocation vs 10k+10k allocations.  Except that the initial 10k allocations are way more than 10k allocations because each struct has nested objects.  so it's really n*10k+1 allocations vs n*10k+10k allocations.  So the additional pointer allocations becoming increasingly irrelevant as the complexity of your base object goes up.

> You're moving costs around (from sort to gc) though by switching to the pointer. I'm just pointing out that I would usually weigh allocation costs and patterns that encourage that >> sort cost (a one time thing in most flows). For anything that is not sorted, it's a no brainer.

there's something to be said for optimizing the user experience which means optimizing operations users wait for over optimizing things that happen in the background (like non-stop-the-world GC, if/when Go gets such a thing)

> Put another way, how expensive is it for you to build a []*Build from a []Build (one allocate and one traversal) and the sort is then faster, so we both win. The converse is not true.

I was pretty sure that basically *was* what we were doing/considering.  (get back list of builds from api call, construct array of pointers to build objects, invoke sort on array of pointers)

**Comment by smarterclayton** [comment]
On Aug 9, 2015, at 10:19 PM, Ben Parees <notifications@github.com> wrote:

I'm talking about normal use. 10k calls to allocate 100 bytes creates 10k
items in GC vs one allocation of 1M is hard to argue is the right approach.

but it's not one allocation in the other case because you still have to
allocate the build objects you retrieved,


I don't know what you mean here.  If you do something like "foo :=
make([]Build, 0, 1000000)" and then pass &foo[1] into DecodeInto, there is
only one allocation (the array).

so you're talking about 10k+1 allocation vs 10k+10k allocations. Except
that the initial 10k allocations are way more than 10k allocations because
each struct has nested objects. so it's really n*10k+1 allocations vs n*10k+10k
allocations. So the additional pointer allocations becoming increasingly
irrelevant as the complexity of your base object goes up.


Only for slices, pointers to structs, and strings, and strings are
allocated during json decode.  I agree in practice n is between 3 and 7.
But the base object is probably the largest of those by several times.


You're moving costs around (from sort to gc) though by switching to the
pointer. I'm just pointing out that I would usually weigh allocation costs
and patterns that encourage that >> sort cost (a one time thing in most
flows). For anything that is not sorted, it's a no brainer.

there's something to be said for optimizing the user experience which means
optimizing operations users wait for over optimizing things that happen in
the background (like non-stop-the-world GC, if/when Go gets such a thing)


Again, I doubt sort will dominate the running time of any practical server
op, but memory and decode will.  Not sure that sort is a significant slice
of that workload to optimize at any of the scales we have (if clients are
sorting 10k items the sort time is going to be drowned by other times, such
that the optimization is of minor impact).

Put another way, how expensive is it for you to build a []*Build from a
[]Build (one allocate and one traversal) and the sort is then faster, so we
both win. The converse is not true.

I was pretty sure that basically *was* what we were doing/considering. (get
back list of builds from api call, construct array of pointers to build
objects, invoke sort on array of pointers)


And that's fine.

—
Reply to this email directly or view it on GitHub
<https://github.com/openshift/origin/pull/4043#issuecomment-129279843>.

**Comment by rhcarvalho** [comment]
@smarterclayton @bparees I wrote my own benchmark based on Ben's:
https://gist.github.com/rhcarvalho/abb12fa958f3e83537cf

Sorting `[]Blob` and `[]*Blob` is about an order of magnitude faster than performing a conversion followed by sort.

Based on that, since we already use `[]Build` and `[]*Build` (converted from `[]Build` in the `pkg/build/prune` package), seems best to have `sort.Interface` implemented on both types.

I've updated this PR in that direction. PTAL.

**Comment by bparees** [comment]
if you want to do efficient conversion, i wouldn't do this:
```
for i := range ps {
			s = append(s, *ps[i])
		}
```
i'd allocate the right sized array and then populate it.  here you're allocating len(ps) array objects.

but i think we've collectively spent more than enough time on this, so i'm fine with the implementation :)

**Comment by rhcarvalho** [comment]
@bparees I've updated the [gist](https://gist.github.com/rhcarvalho/abb12fa958f3e83537cf), even if we preallocate the slice, the conclusion does not change, no conversion is still an order of magnitude faster :-)

```
s := make([]Blob, len(ps))
for i := range ps {
	s[i] = *ps[i]
}
```

**Comment by bparees** [comment]
@rhcarvalho that's definitely surprising... getting the address and assigning it into an array should not be that expensive...

**Comment by rhcarvalho** [comment]
Sorry, "order of magnitude" is actually wrong... my eyes lied to me.
The benchmarks that include converting from values to pointers only make the sorts 1.2~2.2 times slower. Things are a bit worse with converting from pointer to value, making it 3.8~14 times slower.

**Comment by rhcarvalho** [comment]
Can we merge this?

**Comment by bparees** [comment]
[merge]

**Pull request merged**